lib/src/front_end/ast_node_visitor.dart

```diff
@@ -425,37 +426,132 @@ class AstNodeVisitor extends ThrowingAstVisitor<void>
+    //
+    // ```
+    // for (
+    //   initializerClause;
+    //   conditionClause;
+    //   incrementClause
+    // ) {
+    //   body;
+    // }
+    // ```
+    var partsList =
+        DelimitedListBuilder(this, const ListStyle(commas: Commas.none));
+    partsList.leftBracket(node.leftParenthesis);
+
+    // Edge case: A totally empty for loop is formatted just as `(;;)` with
+    // no splits or spaces anywhere.
+    var forLoopParts = node.forLoopParts;
+    if (forLoopParts is ForPartsWithExpression &&
+        forLoopParts.initialization == null &&
+        forLoopParts.leftSeparator.precedingComments == null &&
+        forLoopParts.condition == null &&
+        forLoopParts.rightSeparator.precedingComments == null &&
+        forLoopParts.updaters.isEmpty &&
+        node.rightParenthesis.precedingComments == null) {
+      token(forLoopParts.leftSeparator);
+      token(forLoopParts.rightSeparator);
+    } else {
+      switch (forLoopParts) {
+        // ForParts is the base class for all three-clause "C-style" for loops.
+        // There are three subclasses based on whether the first clause is a
+        // declaration, expression, or pattern. The other clauses are the same
+        // and are handled here.
+        case ForParts():
+          // The initializer clause depends on which subclass we have.
+          switch (forLoopParts) {
+            case ForPartsWithDeclarations():
+              partsList.addCommentsBefore(forLoopParts.variables.beginToken);
+              visit(forLoopParts.variables);
+
+            case ForPartsWithExpression():
+              if (forLoopParts.initialization case var initializer?) {
+                partsList.addCommentsBefore(initializer.beginToken);
+                visit(initializer);
+              } else {
+                partsList.addCommentsBefore(forLoopParts.leftSeparator);
+              }
+
+            case ForPartsWithPattern():
+              throw UnimplementedError();
+          }
+
+          token(forLoopParts.leftSeparator);
+          writer.split();
+          partsList.add(writer.pop());
+
+          // The condition clause.
+          if (forLoopParts.condition case var conditionExpression?) {
+            partsList.addCommentsBefore(conditionExpression.beginToken);
+            visit(conditionExpression);
+          } else {
+            partsList.addCommentsBefore(forLoopParts.rightSeparator);
+          }
+
+          token(forLoopParts.rightSeparator);
+          writer.split();
+          partsList.add(writer.pop());
+
+          // The update clauses.
+          if (forLoopParts.updaters.isNotEmpty) {
+            partsList.addCommentsBefore(forLoopParts.updaters.first.beginToken);
+
+            createList(forLoopParts.updaters,
+                style: const ListStyle(commas: Commas.nonTrailing));
+            partsList.add(writer.pop());
+          }
+
+        case ForEachParts():
+          throw UnimplementedError();
+      }
+    }
+
```

Comment:
[optional] FWIW I do personally like the pattern/switch maximizing version of this code. Moving as much as possible to switch patterns instead of if/else I get:

```dart
    switch (node.forLoopParts) {
      // Edge case: A totally empty for loop is formatted just as `(;;)` with
      // no splits or spaces anywhere.
      case ForPartsWithExpression(
            initialization: null,
            leftSeparator: Token(precedingComments: null) && var leftSeparator,
            condition: null,
            rightSeparator:
                Token(precedingComments: null) && var rightSeparator,
            updaters: NodeList(isEmpty: true),
          )
          when node.rightParenthesis.precedingComments == null:
        token(leftSeparator);
        token(rightSeparator);
      // ForParts is the base class for all three-clause "C-style" for loops.
      // There are three subclasses based on whether the first clause is a
      // declaration, expression, or pattern. The other clauses are the same
      // and are handled here.
      case ForParts() && var forParts:
        // The initializer clause depends on which subclass we have.
        switch (forParts) {
          case ForPartsWithDeclarations(:var variables):
            partsList.addCommentsBefore(variables.beginToken);
            visit(variables);

          case ForPartsWithExpression(:var initialization?):
            partsList.addCommentsBefore(initialization.beginToken);
            visit(initialization);

          case ForPartsWithExpression(:var leftSeparator):
            partsList.addCommentsBefore(leftSeparator);

          case ForPartsWithPattern():
            throw UnimplementedError();
        }

        token(forParts.leftSeparator);
        writer.split();
        partsList.add(writer.pop());

        // The condition clause.
        if (forParts.condition case var conditionExpression?) {
          partsList.addCommentsBefore(conditionExpression.beginToken);
          visit(conditionExpression);
        } else {
          partsList.addCommentsBefore(forParts.rightSeparator);
        }

        token(forParts.rightSeparator);
        writer.split();
        partsList.add(writer.pop());

        // The update clauses.
        if (forParts.updaters.isNotEmpty) {
          partsList.addCommentsBefore(forParts.updaters.first.beginToken);

          createList(forParts.updaters,
              style: const ListStyle(commas: Commas.nonTrailing));
          partsList.add(writer.pop());
        }

      case ForEachParts():
        throw UnimplementedError();
    }
```

And I find that quite readable.

Reply:
Yeah, I'm convinced. I took this approach and also moved the condition/updater part to another function which let me flatten out the nested switches.